### ISSUE TITLE
Add checks for OpenJDK Style chapter - Wrapping Method Declarations

### ISSUE THREAD

**Issue by lithops-zty** [checkstyle/checkstyle#20312]
Parent issue: #19798

<img width="805" height="494" alt="Image" src="https://github.com/user-attachments/assets/2df4ebe7-ccb1-4d3f-843d-4e8a1e054913" />

## Description
Add Checkstyle checks for rules under §3.8 - Wrapping Lines (Wrapping Method Declarations) of the OpenJDK Java Style Guidelines v6.

## Rules Covered

From [§3.8.1 - Wrapping Method Declarations](https://checkstyle.org/styleguides/openjdk-java-style-v6/openjdk-styleguide.html#s3.8.2-wrapping-method-declarations):
1. **Parameter wrapping:** Method arguments can be formatted by listing the arguments vertically, or by a new line and +8 extra spaces.
    - `Indentation Check`, Already covered by #19619 
However there is a false positive when arguments are aligned vertically and indentation is less than +8, for example:
```java
    int f(int a,
          int b) {  // violation ''int' has incorrect indentation level 10'
        ... 
    }
```

2. **Throws clause wrapping:** If a throws clause needs to be wrapped, put the line break in front of the throws clause and make sure it stands out from the argument list, either by indenting +8 relative to the function declaration, or +8 relative to the previous line.
     - New check that validates `throws` is on its own line when the parameter list wraps, and that its indentation is distinct from parameter alignment

3. (from example) **One parameter per line**: If aligning the parameters vertically, don't put two parameters on one line
     -  New check that validates no more than one parameter on the same line when parameters are vertically aligned
     
## Checks to add to `openjdk_checks.xml`

- two new checks for rules 2 and 3

## Progress Tracker

| Check | Rule | PR |
|-------|------|-----|
| `Indentation` | Method argument formatting: vertical or +8 continuation | - |
| (new) | Throws clause on separate line, distinct indentation | |
| (new) | One parameter per line | |

**Comment by Aman-Baliyan** [comment]
@lithops-zty we definitely needs a check for -

`If a throws clause needs to be wrapped, put the line break in front of the throws clause and make sure it stands out from the argument list, either by indenting +8 relative to the function declaration, or +8 relative to the previous line.`

but for this rule -

`Method declarations can be formatted by listing the arguments vertically, or by a new line and +8 extra spaces`

first we have to resolve the false positive.

To resolve the false positive we can add a new property in Identation module which can skip the violation if the method parameter are vertically aligned.

**Comment by lithops-zty** [comment]
@Aman-Baliyan I drafted a plan for the new property in Indentation:

Add new property `allowVerticalAlignment`, default to false.

When true, parameter definitions won't be checked for indentation if **all** method parameters are vertically aligned.

**Strictly vertically aligned:**
```java
    void p(int a,
           int b,
           int c) { }
```
One single misalignment, and the whole method declaration will be subject to check:
**Not strictly vertically aligned:**
```  java
    void p(int a,
           int b,  // violation '11, expected level should be 12.'
          int c) { }  // violation '10, expected level should be 12.'
```

Cases where the first parameter is not on the first line will not be exempted:
```java
    void p(
           int a,  // violation '11, expected level should be 12.'
           int b,  // violation '11, expected level should be 12.'
           int c) { }  // violation '11, expected level should be 12.'
```

Is this design reasonable? Any advice is appreciated!

**Comment by Aman-Baliyan** [comment]
@lithops-zty  design looks good to me , i think this rule is a sub-part of variation3 of section https://cr.openjdk.org/~alundblad/styleguide/index-v6.html#toc-wrapping-lines

<img width="1242" height="180" alt="Image" src="https://github.com/user-attachments/assets/73f84136-e974-4c89-bbf8-594084567ff8" />

is it possible to cover  the variant 3 and variant 4 by this property because there can be same false positives (as your example of method parameters) for constructors argument and chained method call where they are aligned vertically but identation level is less than 8

**Comment by lithops-zty** [comment]
@Aman-Baliyan To me the two variants seems to be two different cases, so I propose to add two properties, one for each variant. (allowVerticalAlignment and allowChainedAlignment)

However, if we extend allowVerticalAlignment beyond method declaration, I am inclined to change the design:

When true, parameter definitions won't be checked for indentation, **as long as each definition vertically aligns with the previous.**

**Strictly vertically aligned:**
```java
    void p(int a,
           int b,
           int c) { }
```
One single misalignment **does not affect other lines:**
**Not strictly vertically aligned:**
```  java
    void p(int a,
           int b,
          int c) { }  // violation '10, expected level should be 12.'
```

Cases where the first parameter is not on the first line: **only the first parameter will be checked.**
```java
    void p(
           int a,  // violation '11, expected level should be 12'
           int b,
           int c) { }
```

This way in implementation we can just check line by line and no special treatment is needed for method declaration.

**Comment by Aman-Baliyan** [comment]
> This way in implementation we can just check line by line and no special treatment is needed for method declaration.

@lithops-zty  While line-by-line validation works cleanly for method declarations, complex wrapped expressions introduce structural variations that are harder to track. 

For example, in a wrapped mathematical expression or assignment:

```c
int a = b + c * d
        - e;
```
if there is a way in `Identation` module to find the start of line-wrapping then your idea of covering variation 3 is possible but if not then we can only cover it for specific tokens like method_def, Constructor_def.

**Comment by Aman-Baliyan** [comment]
@lithops-zty are u working on this issue

**Comment by lithops-zty** [comment]
@Aman-Baliyan Sorry for replying late, was occupied with other commitments.
I am collating a list of all possible scenarios of vertical alignments. There is likely no general method to find the vertical alignment column, but maybe we can try to split these scenarios into groups, each of which can share a common algorithm.

**Comment by Aman-Baliyan** [comment]
> [@Aman-Baliyan](https://github.com/Aman-Baliyan) Sorry for replying late, was occupied with other commitments. I am collating a list of all possible scenarios of vertical alignments. There is likely no general method to find the vertical alignment column, but maybe we can try to split these scenarios into groups, each of which can share a common algorithm.

@lithops-zty No worries you can take your time

**Comment by lithops-zty** [comment]
@Aman-Baliyan Hi, here is my (tentative) list of different scenarios:

1. method/constructor/lambda declaration parameters
1. method/constructor/new call arguments
1. annotation member value pairs
1. arithmetic+boolean expressions
1. ternary expressions
1. chained calls/fields
1. comma-separated variable declarations
1. array initializer with wrapped elements
1. for loop initializer, condition and iterator
1. try resources

I tried and managed to implement logic for all the scenarios: #20440. I'd be happy to listen to your advice!

There are some known issues of the code:
1. A performance bottleneck due to frequent tree traversal in `getTopLeftmostNodeInTree`. We could implement caching in `VerticalAlignmentHandler`, storing the top-leftmost node for each node after computation.

The code is written with reference to this [implementation plan](https://gist.github.com/lithops-zty/d57a024a24bce00b1b2f51728b3e2e39)

**Comment by Aman-Baliyan** [comment]
@romani @mohitsatr please look at this issue

**Comment by romani** [comment]
>Throws clause wrapping

Needs special new Check.
Indentation Check should not be updated as it is already overcomplicated.

**Comment by romani** [comment]
> One parameter per line

We discussed this at some other issue. I shared there reasons why it should be separate Check also.
@lithops-zty , please try to find such discussion.

**Comment by Aman-Baliyan** [comment]
Then what should we do for this false positive case ->


    void p(
           int a,  // violation '11, expected level should be 12'
           int b,
           int c) { }

According to line wrapping rule of Identation there is a violation but if the parameters are vertically aligned then there will be no violation for identation

**Comment by lithops-zty** [comment]
Just to clarify, the example given above is commented with the **intended** violation (i.e. one, at `int a`.), as per OpenJDK Guide.

However, the current `IndentationCheck` configured will incorrectly flag all three parameters:

```java
void p(
       int a,  // violation '11, expected level should be 12'
       int b,  // violation '11, expected level should be 12'
       int c) { }  // violation '11, expected level should be 12'
```

---

Two indentation choices that conforms to OpenJDK Guide:
**All wrapped params indented by 8**
```java
void p(int a,
        int b,
        int c) { }
```

**First param not wrapped, remaining params vertically aligns with first:**
```java
void p(int a,
       int b,
       int c) { }
```

**Comment by lithops-zty** [comment]
> We discussed this at some other issue. I shared there reasons why it should be separate Check also.
> @lithops-zty , please try to find such discussion.

@romani I found it, #9118. I agree that the `one param per line` and `throw wrapping` rule should be enforced by new checks. But as @Aman-Baliyan has pointed out, `IndentationCheck` emits false positives which must be addressed first.

I also see it is a very long discussion with no eventual consensus on design. There were 2 designs implemented but neither was added to sevntu-checkstyle and time-proven. I don't think I can propose any better alternatives than those well-discussed ones.

**Comment by Aman-Baliyan** [comment]
> Then what should we do for this false positive case ->
> 
> ```
> void p(
>        int a,  // violation '11, expected level should be 12'
>        int b,
>        int c) { }
> ```
> 
> According to line wrapping rule of Identation there is a violation but if the parameters are vertically aligned then there will be no violation for identation

@romani  should we add suppression for method parameters for now

**Comment by romani** [comment]
Yes, suppression for now

**Comment by Aman-Baliyan** [comment]
I think my above assumption is wrong for false positive because in real life codebases there will not present any method like this 

```
void p(
       int a,  // violation '11, expected level should be 12'
       int b,
       int c) { }
```
So there will be no need of suppression.

**Comment by romani** [comment]
very rare way to write method, we can improve later on users request.

**Comment by Aman-Baliyan** [comment]
> very rare way to write method, we can improve later on users request.

Then this issue is done. We can close it
